From eef8c4c2da777be81f02e860ad2a0f0d46c5ab6b Mon Sep 17 00:00:00 2001
From: "Arun P. Mohanan" <arun.p.m@linux.intel.com>
Date: Wed, 27 Apr 2022 10:16:47 +0530
Subject: [PATCH 1/5] i2c: mux: Add mux deselect support on timeout

Add support to deselect the mux when there is a timeout.
The mux idle_state settings will be configured on startup. In case of
MCTP it is MUX_IDLE_DISCONNECT. But when there is a timeout, mux ends
up in connected position and the devices behind the mux will appear under
different muxes connected to the same bus. This change fix the same.

Signed-off-by: Arun P. Mohanan <arun.p.m@linux.intel.com>
---
 drivers/i2c/i2c-mux.c | 62 +++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 62 insertions(+)

diff --git a/drivers/i2c/i2c-mux.c b/drivers/i2c/i2c-mux.c
index 3f1302bd0c23..d0e6cb6fa07d 100644
--- a/drivers/i2c/i2c-mux.c
+++ b/drivers/i2c/i2c-mux.c
@@ -36,6 +36,32 @@ struct i2c_mux_priv {
 	u32 chan_id;
 };
 
+static void i2c_mux_hold(struct i2c_mux_core *muxc, u32 chan_id,
+			 unsigned long timeout)
+{
+	mutex_lock(&muxc->hold_lock);
+	muxc->holder_chan_id = chan_id;
+	schedule_delayed_work(&muxc->unhold_work, timeout);
+}
+
+static void i2c_mux_unhold(struct i2c_mux_core *muxc)
+{
+	cancel_delayed_work_sync(&muxc->unhold_work);
+	mutex_unlock(&muxc->hold_lock);
+}
+
+static void i2c_mux_unhold_work(struct work_struct *work)
+{
+	struct delayed_work *dwork = to_delayed_work(work);
+	struct i2c_mux_core *muxc = container_of(dwork, struct i2c_mux_core,
+						 unhold_work);
+
+	if (muxc->deselect)
+		muxc->deselect(muxc, muxc->holder_chan_id);
+
+	mutex_unlock(&muxc->hold_lock);
+}
+
 static int __i2c_mux_master_xfer(struct i2c_adapter *adap,
 				 struct i2c_msg msgs[], int num)
 {
@@ -46,6 +72,15 @@ static int __i2c_mux_master_xfer(struct i2c_adapter *adap,
 
 	/* Switch to the right mux port and perform the transfer. */
 
+	hold_msg = i2c_check_hold_msg(msgs[num - 1].flags,
+				      msgs[num - 1].len,
+				      (u16 *)msgs[num - 1].buf);
+	if (hold_msg == I2C_HOLD_MSG_SET) {
+		timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
+	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+		mutex_lock(&muxc->hold_lock);
+	}
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = __i2c_transfer(parent, msgs, num);
@@ -65,6 +100,15 @@ static int i2c_mux_master_xfer(struct i2c_adapter *adap,
 
 	/* Switch to the right mux port and perform the transfer. */
 
+	hold_msg = i2c_check_hold_msg(msgs[num - 1].flags,
+				      msgs[num - 1].len,
+				      (u16 *)msgs[num - 1].buf);
+	if (hold_msg == I2C_HOLD_MSG_SET) {
+		timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
+	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+		mutex_lock(&muxc->hold_lock);
+	}
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = i2c_transfer(parent, msgs, num);
@@ -86,6 +130,15 @@ static int __i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 
 	/* Select the right mux port and perform the transfer. */
 
+	hold_msg = i2c_check_hold_msg(flags,
+				      size == I2C_SMBUS_WORD_DATA ? 2 : 0,
+				      &data->word);
+	if (hold_msg == I2C_HOLD_MSG_SET) {
+		timeout = msecs_to_jiffies(data->word);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
+	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+		mutex_lock(&muxc->hold_lock);
+	}
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = __i2c_smbus_xfer(parent, addr, flags,
@@ -108,6 +161,15 @@ static int i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 
 	/* Select the right mux port and perform the transfer. */
 
+	hold_msg = i2c_check_hold_msg(flags,
+				      size == I2C_SMBUS_WORD_DATA ? 2 : 0,
+				      &data->word);
+	if (hold_msg == I2C_HOLD_MSG_SET) {
+		timeout = msecs_to_jiffies(data->word);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
+	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+		mutex_lock(&muxc->hold_lock);
+	}
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = i2c_smbus_xfer(parent, addr, flags,
-- 
2.41.0


From 37cf45d5373a27624c8292c11de215d13af91f8b Mon Sep 17 00:00:00 2001
From: Jae Hyun Yoo <jae.hyun.yoo@intel.com>
Date: Fri, 15 Feb 2019 16:05:09 -0800
Subject: [PATCH 2/5] i2c: Add mux hold/unhold msg types

This commit adds mux hold/unhold message types to support extended
mux control for IPMB and MCTP devices. A hold or an unhold message
can be added at the end of I2C message stream wrapped by
repeated-start, also can be used as a single message independantly.

This mux hold/unhold message will be delivered throughout all mux
levels in the path. Means that if it goes to multi-level mux path,
all muxes will be held/unheld by this message.

1. Hold message
   struct i2c_msg msg;
   uint16_t timeout = 5000; // timeout in ms. 5 secs in this example.

   msg.addr = 0x0; // any value can be used. addr will be ignored in this packet.
   msg.flags = I2C_M_HOLD; // set this flag to indicate it's a hold message.
   msg.len = sizeof(uint16_t); // timeout value will be delivered using two bytes buffer.
   msg.buf = (uint8_t *)&timeout; // set timeout value.

2. Unhold message
   struct i2c_msg msg;
   uint16_t timeout = 0; // set 0 for an unhold message.

   msg.addr = 0x0; // any value can be used. addr will be ignored in this packet.
   msg.flags = I2C_M_HOLD; // set this flag to indicate it's an unhold message.
   msg.len = sizeof(uint16_t); // timeout value will be delivered using two bytes buffer.
   msg.buf = (uint8_t *)&timeout; // set timeout value.

   This unhold message can be delivered to a mux adapter even when
   a bus is locked so that any holding state can be unheld
   immediately by invoking this unhold message.

This patch would not be welcomed from upstream so it should be kept
in downstream only.

Signed-off-by: Jae Hyun Yoo <jae.hyun.yoo@intel.com>
---
 drivers/i2c/i2c-core-base.c  | 76 ++++++++++++++++++++++++++++++++++--
 drivers/i2c/i2c-core-smbus.c | 22 +++++++++--
 drivers/i2c/i2c-mux.c        | 70 ++++++++++++++++++++++++---------
 3 files changed, 142 insertions(+), 26 deletions(-)

diff --git a/drivers/i2c/i2c-core-base.c b/drivers/i2c/i2c-core-base.c
index 096001225dc4..2d8f4f9ebec7 100644
--- a/drivers/i2c/i2c-core-base.c
+++ b/drivers/i2c/i2c-core-base.c
@@ -1509,6 +1509,27 @@ int i2c_handle_smbus_host_notify(struct i2c_adapter *adap, unsigned short addr)
 }
 EXPORT_SYMBOL_GPL(i2c_handle_smbus_host_notify);
 
+static void i2c_adapter_hold(struct i2c_adapter *adapter, unsigned long timeout)
+{
+	mutex_lock(&adapter->hold_lock);
+	schedule_delayed_work(&adapter->unhold_work, timeout);
+}
+
+static void i2c_adapter_unhold(struct i2c_adapter *adapter)
+{
+	cancel_delayed_work_sync(&adapter->unhold_work);
+	mutex_unlock(&adapter->hold_lock);
+}
+
+static void i2c_adapter_unhold_work(struct work_struct *work)
+{
+	struct delayed_work *dwork = to_delayed_work(work);
+	struct i2c_adapter *adapter = container_of(dwork, struct i2c_adapter,
+						   unhold_work);
+
+	mutex_unlock(&adapter->hold_lock);
+}
+
 static int i2c_register_adapter(struct i2c_adapter *adap)
 {
 	int res = -EINVAL;
@@ -1597,6 +1618,9 @@ static int i2c_register_adapter(struct i2c_adapter *adap)
 	bus_for_each_drv(&i2c_bus_type, NULL, adap, __process_new_adapter);
 	mutex_unlock(&core_lock);
 
+	mutex_init(&adap->hold_lock);
+	INIT_DELAYED_WORK(&adap->unhold_work, i2c_adapter_unhold_work);
+
 	return 0;
 
 out_reg:
@@ -1817,6 +1841,8 @@ void i2c_del_adapter(struct i2c_adapter *adap)
 	idr_remove(&i2c_adapter_idr, adap->nr);
 	mutex_unlock(&core_lock);
 
+	i2c_adapter_unhold(adap);
+
 	/* Clear the device structure in case this adapter is ever going to be
 	   added again */
 	memset(&adap->dev, 0, sizeof(adap->dev));
@@ -2230,7 +2256,9 @@ static int i2c_check_for_quirks(struct i2c_adapter *adap, struct i2c_msg *msgs,
  */
 int __i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 {
+	enum i2c_hold_msg_type hold_msg = I2C_HOLD_MSG_NONE;
 	unsigned long orig_jiffies;
+	unsigned long timeout;
 	int ret, try;
 
 	if (!adap->algo->master_xfer) {
@@ -2248,6 +2276,25 @@ int __i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 	if (adap->quirks && i2c_check_for_quirks(adap, msgs, num))
 		return -EOPNOTSUPP;
 
+	/* Do not deliver a mux hold msg to root bus adapter */
+	if (!i2c_parent_is_i2c_adapter(adap)) {
+	    hold_msg = i2c_check_hold_msg(msgs[num - 1].flags,
+					   msgs[num - 1].len,
+					  (u16 *)msgs[num - 1].buf);
+		if (hold_msg == I2C_HOLD_MSG_SET) {
+			timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
+			i2c_adapter_hold(adap, timeout);
+
+			if (--num == 0)
+				return 0;
+		} else if (hold_msg == I2C_HOLD_MSG_RESET) {
+			i2c_adapter_unhold(adap);
+			return 0;
+		} else if (hold_msg == I2C_HOLD_MSG_NONE) {
+			mutex_lock(&adap->hold_lock);
+		}
+	}
+
 	/*
 	 * i2c_trace_msg_key gets enabled when tracepoint i2c_transfer gets
 	 * enabled.  This is an efficient way of keeping the for-loop from
@@ -2284,6 +2331,13 @@ int __i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 		trace_i2c_result(adap, num, ret);
 	}
 
+	if (!i2c_parent_is_i2c_adapter(adap)) {
+		if (hold_msg == I2C_HOLD_MSG_SET && ret < 0)
+			i2c_adapter_unhold(adap);
+		else if (hold_msg == I2C_HOLD_MSG_NONE)
+			mutex_unlock(&adap->hold_lock);
+	}
+
 	return ret;
 }
 EXPORT_SYMBOL(__i2c_transfer);
@@ -2302,6 +2356,7 @@ EXPORT_SYMBOL(__i2c_transfer);
  */
 int i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 {
+	bool do_bus_lock = true;
 	int ret;
 
 	/* REVISIT the fault reporting model here is weak:
@@ -2320,12 +2375,25 @@ int i2c_transfer(struct i2c_adapter *adap, struct i2c_msg *msgs, int num)
 	 *    one (discarding status on the second message) or errno
 	 *    (discarding status on the first one).
 	 */
-	ret = __i2c_lock_bus_helper(adap);
-	if (ret)
-		return ret;
+	/*
+	 * Do not lock a bus for delivering an unhold msg to a mux
+	 * adpater. This is just for a single length unhold msg case.
+	 */
+	if (num == 1 && i2c_parent_is_i2c_adapter(adap) &&
+	    i2c_check_hold_msg(msgs[0].flags, msgs[0].len,
+			       (u16 *)msgs[0].buf) ==
+			       I2C_HOLD_MSG_RESET)
+		do_bus_lock = false;
+
+	if (do_bus_lock) {
+		ret = __i2c_lock_bus_helper(adap);
+		if (ret)
+			return ret;
+	}
 
 	ret = __i2c_transfer(adap, msgs, num);
-	i2c_unlock_bus(adap, I2C_LOCK_SEGMENT);
+	if (do_bus_lock)
+		i2c_unlock_bus(adap, I2C_LOCK_SEGMENT);
 
 	return ret;
 }
diff --git a/drivers/i2c/i2c-core-smbus.c b/drivers/i2c/i2c-core-smbus.c
index e3b96fc53b5c..f5b9615d4b65 100644
--- a/drivers/i2c/i2c-core-smbus.c
+++ b/drivers/i2c/i2c-core-smbus.c
@@ -536,15 +536,29 @@ s32 i2c_smbus_xfer(struct i2c_adapter *adapter, u16 addr,
 		   unsigned short flags, char read_write,
 		   u8 command, int protocol, union i2c_smbus_data *data)
 {
+	bool do_bus_lock = true;
 	s32 res;
 
-	res = __i2c_lock_bus_helper(adapter);
-	if (res)
-		return res;
+	/*
+	 * Do not lock a bus for delivering an unhold msg to a mux adpater.
+	 * This is just for a single length unhold msg case.
+	 */
+	if (i2c_parent_is_i2c_adapter(adapter) &&
+	    i2c_check_hold_msg(flags,
+			       protocol == I2C_SMBUS_WORD_DATA ? 2 : 0,
+			       &data->word) == I2C_HOLD_MSG_RESET)
+		do_bus_lock = false;
+
+	if (do_bus_lock) {
+		res = __i2c_lock_bus_helper(adapter);
+		if (res)
+			return res;
+	}
 
 	res = __i2c_smbus_xfer(adapter, addr, flags, read_write,
 			       command, protocol, data);
-	i2c_unlock_bus(adapter, I2C_LOCK_SEGMENT);
+	if (do_bus_lock)
+		i2c_unlock_bus(adapter, I2C_LOCK_SEGMENT);
 
 	return res;
 }
diff --git a/drivers/i2c/i2c-mux.c b/drivers/i2c/i2c-mux.c
index d0e6cb6fa07d..41b985e581d1 100644
--- a/drivers/i2c/i2c-mux.c
+++ b/drivers/i2c/i2c-mux.c
@@ -36,11 +36,9 @@ struct i2c_mux_priv {
 	u32 chan_id;
 };
 
-static void i2c_mux_hold(struct i2c_mux_core *muxc, u32 chan_id,
-			 unsigned long timeout)
+static void i2c_mux_hold(struct i2c_mux_core *muxc, unsigned long timeout)
 {
 	mutex_lock(&muxc->hold_lock);
-	muxc->holder_chan_id = chan_id;
 	schedule_delayed_work(&muxc->unhold_work, timeout);
 }
 
@@ -56,9 +54,6 @@ static void i2c_mux_unhold_work(struct work_struct *work)
 	struct i2c_mux_core *muxc = container_of(dwork, struct i2c_mux_core,
 						 unhold_work);
 
-	if (muxc->deselect)
-		muxc->deselect(muxc, muxc->holder_chan_id);
-
 	mutex_unlock(&muxc->hold_lock);
 }
 
@@ -68,6 +63,8 @@ static int __i2c_mux_master_xfer(struct i2c_adapter *adap,
 	struct i2c_mux_priv *priv = adap->algo_data;
 	struct i2c_mux_core *muxc = priv->muxc;
 	struct i2c_adapter *parent = muxc->parent;
+	enum i2c_hold_msg_type hold_msg;
+	unsigned long timeout;
 	int ret;
 
 	/* Switch to the right mux port and perform the transfer. */
@@ -77,15 +74,23 @@ static int __i2c_mux_master_xfer(struct i2c_adapter *adap,
 				      (u16 *)msgs[num - 1].buf);
 	if (hold_msg == I2C_HOLD_MSG_SET) {
 		timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
-		i2c_mux_hold(muxc, priv->chan_id, timeout);
+		i2c_mux_hold(muxc, timeout);
 	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
 		mutex_lock(&muxc->hold_lock);
 	}
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = __i2c_transfer(parent, msgs, num);
-	if (muxc->deselect)
-		muxc->deselect(muxc, priv->chan_id);
+	if (hold_msg != I2C_HOLD_MSG_SET) {
+		if (muxc->deselect)
+			muxc->deselect(muxc, priv->chan_id);
+		if (hold_msg == I2C_HOLD_MSG_RESET)
+			i2c_mux_unhold(muxc);
+		else
+			mutex_unlock(&muxc->hold_lock);
+	} else if (hold_msg == I2C_HOLD_MSG_SET && ret < 0) {
+		i2c_mux_unhold(muxc);
+	}
 
 	return ret;
 }
@@ -96,6 +101,8 @@ static int i2c_mux_master_xfer(struct i2c_adapter *adap,
 	struct i2c_mux_priv *priv = adap->algo_data;
 	struct i2c_mux_core *muxc = priv->muxc;
 	struct i2c_adapter *parent = muxc->parent;
+	enum i2c_hold_msg_type hold_msg;
+	unsigned long timeout;
 	int ret;
 
 	/* Switch to the right mux port and perform the transfer. */
@@ -105,15 +112,21 @@ static int i2c_mux_master_xfer(struct i2c_adapter *adap,
 				      (u16 *)msgs[num - 1].buf);
 	if (hold_msg == I2C_HOLD_MSG_SET) {
 		timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
-		i2c_mux_hold(muxc, priv->chan_id, timeout);
+		i2c_mux_hold(muxc, timeout);
 	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
 		mutex_lock(&muxc->hold_lock);
 	}
 	ret = muxc->select(muxc, priv->chan_id);
 	if (ret >= 0)
 		ret = i2c_transfer(parent, msgs, num);
-	if (muxc->deselect)
-		muxc->deselect(muxc, priv->chan_id);
+	if (hold_msg != I2C_HOLD_MSG_SET) {
+		if (muxc->deselect)
+			muxc->deselect(muxc, priv->chan_id);
+		if (hold_msg == I2C_HOLD_MSG_RESET)
+			i2c_mux_unhold(muxc);
+		else
+			mutex_unlock(&muxc->hold_lock);
+	}
 
 	return ret;
 }
@@ -126,6 +139,8 @@ static int __i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 	struct i2c_mux_priv *priv = adap->algo_data;
 	struct i2c_mux_core *muxc = priv->muxc;
 	struct i2c_adapter *parent = muxc->parent;
+	enum i2c_hold_msg_type hold_msg;
+	unsigned long timeout;
 	int ret;
 
 	/* Select the right mux port and perform the transfer. */
@@ -135,7 +150,7 @@ static int __i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 				      &data->word);
 	if (hold_msg == I2C_HOLD_MSG_SET) {
 		timeout = msecs_to_jiffies(data->word);
-		i2c_mux_hold(muxc, priv->chan_id, timeout);
+		i2c_mux_hold(muxc, timeout);
 	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
 		mutex_lock(&muxc->hold_lock);
 	}
@@ -143,8 +158,14 @@ static int __i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 	if (ret >= 0)
 		ret = __i2c_smbus_xfer(parent, addr, flags,
 				       read_write, command, size, data);
-	if (muxc->deselect)
-		muxc->deselect(muxc, priv->chan_id);
+	if (hold_msg != I2C_HOLD_MSG_SET) {
+		if (muxc->deselect)
+			muxc->deselect(muxc, priv->chan_id);
+		if (hold_msg == I2C_HOLD_MSG_RESET)
+			i2c_mux_unhold(muxc);
+		else
+			mutex_unlock(&muxc->hold_lock);
+	}
 
 	return ret;
 }
@@ -157,6 +178,8 @@ static int i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 	struct i2c_mux_priv *priv = adap->algo_data;
 	struct i2c_mux_core *muxc = priv->muxc;
 	struct i2c_adapter *parent = muxc->parent;
+	enum i2c_hold_msg_type hold_msg;
+	unsigned long timeout;
 	int ret;
 
 	/* Select the right mux port and perform the transfer. */
@@ -166,7 +189,7 @@ static int i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 				      &data->word);
 	if (hold_msg == I2C_HOLD_MSG_SET) {
 		timeout = msecs_to_jiffies(data->word);
-		i2c_mux_hold(muxc, priv->chan_id, timeout);
+		i2c_mux_hold(muxc, timeout);
 	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
 		mutex_lock(&muxc->hold_lock);
 	}
@@ -174,8 +197,14 @@ static int i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 	if (ret >= 0)
 		ret = i2c_smbus_xfer(parent, addr, flags,
 				     read_write, command, size, data);
-	if (muxc->deselect)
-		muxc->deselect(muxc, priv->chan_id);
+	if (hold_msg != I2C_HOLD_MSG_SET) {
+		if (muxc->deselect)
+			muxc->deselect(muxc, priv->chan_id);
+		if (hold_msg == I2C_HOLD_MSG_RESET)
+			i2c_mux_unhold(muxc);
+		else
+			mutex_unlock(&muxc->hold_lock);
+	}
 
 	return ret;
 }
@@ -369,6 +398,9 @@ struct i2c_mux_core *i2c_mux_alloc(struct i2c_adapter *parent,
 	muxc->deselect = deselect;
 	muxc->max_adapters = max_adapters;
 
+	mutex_init(&muxc->hold_lock);
+	INIT_DELAYED_WORK(&muxc->unhold_work, i2c_mux_unhold_work);
+
 	return muxc;
 }
 EXPORT_SYMBOL_GPL(i2c_mux_alloc);
@@ -554,6 +586,8 @@ void i2c_mux_del_adapters(struct i2c_mux_core *muxc)
 {
 	char symlink_name[20];
 
+	i2c_mux_unhold(muxc);
+
 	while (muxc->num_adapters) {
 		struct i2c_adapter *adap = muxc->adapter[--muxc->num_adapters];
 		struct i2c_mux_priv *priv = adap->algo_data;
-- 
2.41.0


From 7cf9092e327b6d6c5cdf0a753f5bd9b918fa96db Mon Sep 17 00:00:00 2001
From: Radivoje Jovanovic <rjovanovic@nvidia.com>
Date: Tue, 11 Oct 2022 12:27:14 -0700
Subject: [PATCH 3/5] drivers: i2c: arb add the support for Nvidia chassis

original mr:
https://gitlab-master.nvidia.com/dgx/bmc/linux/-/merge_requests/119

Signed-off-by: Radivoje (Ogi) Jovanovic <rjovanovic@nvidia.com>
fixes JIRA https://jirasw.nvidia.com/browse/DGXOPENBMC-5393
---
 .../dts/aspeed-bmc-nvidia-gh-i2c-arb.dtsi     |  13 ++
 drivers/i2c/i2c-mux.c                         |   6 +-
 drivers/i2c/muxes/Kconfig                     |   6 +-
 drivers/i2c/muxes/i2c-arb-gpio-challenge.c    | 176 ++++++++++++++++++
 4 files changed, 199 insertions(+), 2 deletions(-)
 create mode 100644 arch/arm/boot/dts/aspeed-bmc-nvidia-gh-i2c-arb.dtsi

diff --git a/arch/arm/boot/dts/aspeed-bmc-nvidia-gh-i2c-arb.dtsi b/arch/arm/boot/dts/aspeed-bmc-nvidia-gh-i2c-arb.dtsi
new file mode 100644
index 000000000000..f41d820a4ae1
--- /dev/null
+++ b/arch/arm/boot/dts/aspeed-bmc-nvidia-gh-i2c-arb.dtsi
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: GPL-2.0+
+
+#include "aspeed-g6.dtsi"
+
+/ {
+
+	i2carb: i2carb {
+		compatible = "i2c-arb-gpio-challenge";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "disabled";
+	};
+};
diff --git a/drivers/i2c/i2c-mux.c b/drivers/i2c/i2c-mux.c
index 41b985e581d1..c5290e4062a7 100644
--- a/drivers/i2c/i2c-mux.c
+++ b/drivers/i2c/i2c-mux.c
@@ -531,7 +531,11 @@ int i2c_mux_add_adapter(struct i2c_mux_core *muxc,
 			}
 		}
 
-		priv->adap.dev.of_node = child;
+		if (child)
+			priv->adap.dev.of_node = child;
+		else
+			priv->adap.dev.of_node = mux_node;
+
 		of_node_put(mux_node);
 	}
 
diff --git a/drivers/i2c/muxes/Kconfig b/drivers/i2c/muxes/Kconfig
index db1b9057612a..8bd05dbaea9d 100644
--- a/drivers/i2c/muxes/Kconfig
+++ b/drivers/i2c/muxes/Kconfig
@@ -118,5 +118,9 @@ config I2C_MUX_MLXCPLD
 
 	  This driver can also be built as a module.  If so, the module
 	  will be called i2c-mux-mlxcpld.
-
+config DRAGON_CHASSIS
+	bool "Nvidia Dragon Chassis support"
+	help
+	  This enables all the necessary code for the support of the
+	  Dragon Chassis features.
 endmenu
diff --git a/drivers/i2c/muxes/i2c-arb-gpio-challenge.c b/drivers/i2c/muxes/i2c-arb-gpio-challenge.c
index 24168e9f7df4..3ab81931482c 100644
--- a/drivers/i2c/muxes/i2c-arb-gpio-challenge.c
+++ b/drivers/i2c/muxes/i2c-arb-gpio-challenge.c
@@ -14,6 +14,10 @@
 #include <linux/platform_device.h>
 #include <linux/slab.h>
 
+#if CONFIG_DRAGON_CHASSIS
+#include <linux/device.h>
+#include <linux/timer.h>
+#endif //CONFIG_DRAGON_CHASSIS
 
 /**
  * struct i2c_arbitrator_data - Driver data for I2C arbitrator
@@ -28,9 +32,17 @@
 struct i2c_arbitrator_data {
 	struct gpio_desc *our_gpio;
 	struct gpio_desc *their_gpio;
+	struct gpio_desc *cpld_refresh;
 	unsigned int slew_delay_us;
 	unsigned int wait_retry_us;
 	unsigned int wait_free_us;
+	struct timer_list timer;
+	struct i2c_mux_core *muxc;
+#if CONFIG_DRAGON_CHASSIS
+	struct i2c_adapter *cpld_access;
+	u8 cpld_address;
+	bool bus_locked;
+#endif //CONFIG_DRAGON_CHASSIS
 };
 
 
@@ -44,6 +56,17 @@ static int i2c_arbitrator_select(struct i2c_mux_core *muxc, u32 chan)
 	const struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
 	unsigned long stop_retry, stop_time;
 
+#if CONFIG_DRAGON_CHASSIS
+	if (!IS_ERR(arb->cpld_refresh)) {
+		if (gpiod_get_value(arb->cpld_refresh))
+			return -EBUSY;
+	}
+
+	if (arb->bus_locked)
+		return 0;
+
+#endif //CONFIG_DRAGON_CHASSIS
+
 	/* Start a round of trying to claim the bus */
 	stop_time = jiffies + usecs_to_jiffies(arb->wait_free_us) + 1;
 	do {
@@ -86,6 +109,11 @@ static int i2c_arbitrator_deselect(struct i2c_mux_core *muxc, u32 chan)
 {
 	const struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
 
+#if CONFIG_DRAGON_CHASSIS
+	if (arb->bus_locked) //if bus is locked do not deselect it
+		return 0;
+#endif //CONFIG_DRAGON_CHASSIS
+
 	/* Release the bus and wait for the other master to notice */
 	gpiod_set_value(arb->our_gpio, 0);
 	udelay(arb->slew_delay_us);
@@ -93,6 +121,117 @@ static int i2c_arbitrator_deselect(struct i2c_mux_core *muxc, u32 chan)
 	return 0;
 }
 
+#if CONFIG_DRAGON_CHASSIS
+static ssize_t cpld_refresh_show(struct device *dev,
+				 struct device_attribute *attr,
+				 char *buf)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct i2c_mux_core *muxc = i2c_get_clientdata(client);
+	const struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
+	int gpio_val = -1;
+
+	if (!IS_ERR(arb->cpld_refresh))
+		gpio_val = gpiod_get_value(arb->cpld_refresh);
+
+	return sprintf(buf, "%d\n", gpio_val);
+}
+static DEVICE_ATTR_RO(cpld_refresh);
+
+static ssize_t arb_state_show(struct device *dev,
+			      struct device_attribute *attr,
+			      char *buf)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct i2c_mux_core *muxc = i2c_get_clientdata(client);
+	const struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
+
+	int gpio_val = gpiod_get_value(arb->their_gpio);
+
+	if (!IS_ERR(arb->cpld_refresh)) {
+		if (gpiod_get_value(arb->cpld_refresh))
+			return -EBUSY;
+	}
+
+	return sprintf(buf, "%d\n", gpio_val);
+}
+
+static ssize_t arb_state_store(struct device *dev,
+			       struct device_attribute *attr,
+			       const char *buf, size_t count)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct i2c_mux_core *muxc = i2c_get_clientdata(client);
+	struct i2c_arbitrator_data *arb = i2c_mux_priv(muxc);
+	int val;
+	int ret;
+
+	ret = kstrtoint(buf, 0, &val);
+	if (ret < 0)
+		return ret;
+
+	if (val != 0 && val != 1)
+		return -EINVAL;
+
+	if (!val) {
+		arb->bus_locked = false;
+		del_timer_sync(&arb->timer);
+		i2c_arbitrator_deselect(muxc, 0);
+	} else {
+		ret = i2c_arbitrator_select(muxc, 0);
+		if (ret < 0)
+			return ret;
+		arb->timer.expires = jiffies + HZ;
+		add_timer(&arb->timer);
+		arb->bus_locked = true;
+	}
+	return count;
+}
+static DEVICE_ATTR_RW(arb_state);
+
+static void pet_cpld_register(struct timer_list *t)
+{
+	struct i2c_arbitrator_data *arb = from_timer(arb, t, timer);
+	struct i2c_mux_core *muxc = arb->muxc;
+	int gpio_val;
+	u8 out_buf0[2];
+	u8 out_buf1[2];
+	struct i2c_msg msgs[] = {
+		{
+			.addr = arb->cpld_address,
+			.flags = 0,
+			.len = 2,
+			.buf = out_buf0,
+		},
+		{
+			.addr = arb->cpld_address,
+			.flags = 0,
+			.len = 2,
+			.buf = out_buf1,
+		}
+	};
+
+	out_buf0[0] = 16;
+	out_buf0[1] = 0x5;
+	out_buf1[0] = 16;
+	out_buf1[1] = 0xA;
+
+	gpio_val = gpiod_get_value(arb->their_gpio);
+
+	if (!gpio_val)
+	{
+		if (i2c_transfer(arb->cpld_access, msgs, 2) != 2)
+			dev_err(muxc->dev, "Could not access cpld to pet the watchdog\n");
+
+		mod_timer(&arb->timer, jiffies + HZ);
+	}
+	else
+	{
+			dev_err(muxc->dev, "Unexpected gpio value in timer.\n");
+	}
+}
+#endif //CONFIG_DRAGON_CHASSIS
+
 static int i2c_arbitrator_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
@@ -103,6 +242,10 @@ static int i2c_arbitrator_probe(struct platform_device *pdev)
 	struct gpio_desc *dummy;
 	int ret;
 
+#if CONFIG_DRAGON_CHASSIS
+	struct device_node *cpld_np;
+#endif //CONFIG_DRAGON_CHASSIS
+
 	/* We only support probing from device tree; no platform_data */
 	if (!np) {
 		dev_err(dev, "Cannot find device tree node\n");
@@ -118,6 +261,7 @@ static int i2c_arbitrator_probe(struct platform_device *pdev)
 	if (!muxc)
 		return -ENOMEM;
 	arb = i2c_mux_priv(muxc);
+	arb->muxc = muxc;
 
 	platform_set_drvdata(pdev, muxc);
 
@@ -136,6 +280,8 @@ static int i2c_arbitrator_probe(struct platform_device *pdev)
 		return PTR_ERR(arb->their_gpio);
 	}
 
+	arb->cpld_refresh = devm_gpiod_get(dev, "cpld-refresh", GPIOD_IN);
+
 	/* At the moment we only support a single two master (us + 1 other) */
 	dummy = devm_gpiod_get_index(dev, "their-claim", 1, GPIOD_IN);
 	if (!IS_ERR(dummy)) {
@@ -153,6 +299,24 @@ static int i2c_arbitrator_probe(struct platform_device *pdev)
 	if (of_property_read_u32(np, "wait-free-us", &arb->wait_free_us))
 		arb->wait_free_us = 50000;
 
+#if CONFIG_DRAGON_CHASSIS
+	if (of_property_read_u8(np, "cpld-address", &arb->cpld_address))
+		return -EINVAL;
+
+	/* Find our cpld i2c bus */
+	cpld_np = of_parse_phandle(np, "i2c-cpld", 0);
+	if (!cpld_np) {
+		dev_err(dev, "Cannot parse i2c-cpld\n");
+		return -EINVAL;
+	}
+	arb->cpld_access = of_get_i2c_adapter_by_node(cpld_np);
+	of_node_put(cpld_np);
+	if (!arb->cpld_access) {
+		dev_err(dev, "Cannot find cpld bus\n");
+		return -EPROBE_DEFER;
+	}
+#endif //CONFIG_DRAGON_CHASSIS
+
 	/* Find our parent */
 	parent_np = of_parse_phandle(np, "i2c-parent", 0);
 	if (!parent_np) {
@@ -171,6 +335,13 @@ static int i2c_arbitrator_probe(struct platform_device *pdev)
 	if (ret)
 		i2c_put_adapter(muxc->parent);
 
+#if CONFIG_DRAGON_CHASSIS
+	timer_setup(&arb->timer, pet_cpld_register, 0);
+	device_create_file(dev, &dev_attr_arb_state);
+	device_create_file(dev, &dev_attr_cpld_refresh);
+	arb->bus_locked = false;
+#endif //CONFIG_DRAGON_CHASSIS
+
 	return ret;
 }
 
@@ -178,6 +349,11 @@ static void i2c_arbitrator_remove(struct platform_device *pdev)
 {
 	struct i2c_mux_core *muxc = platform_get_drvdata(pdev);
 
+#if CONFIG_DRAGON_CHASSIS
+	device_remove_file(&pdev->dev, &dev_attr_arb_state);
+	device_remove_file(&pdev->dev, &dev_attr_cpld_refresh);
+#endif //CONFIG_DRAGON_CHASSIS
+
 	i2c_mux_del_adapters(muxc);
 	i2c_put_adapter(muxc->parent);
 }
-- 
2.41.0


From d26d1d0d246cc602152bde8efe96cf8ffd98e3e6 Mon Sep 17 00:00:00 2001
From: Ben Peled <bpeled@nvidia.com>
Date: Tue, 21 Nov 2023 06:15:45 -0800
Subject: [PATCH 4/5] init hold lock before registering devices

some devices register includes i2c transfer
i2c transfer uses hold lock
this creates situation were lock is taken before init

Signed-off-by: Ben Peled <bpeled@nvidia.com>
---
 drivers/i2c/i2c-core-base.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/drivers/i2c/i2c-core-base.c b/drivers/i2c/i2c-core-base.c
index 2d8f4f9ebec7..cbdae572de46 100644
--- a/drivers/i2c/i2c-core-base.c
+++ b/drivers/i2c/i2c-core-base.c
@@ -1604,7 +1604,7 @@ static int i2c_register_adapter(struct i2c_adapter *adap)
 		dev_warn(&adap->dev,
 			 "Failed to create compatibility class link\n");
 #endif
-
+	mutex_init(&adap->hold_lock);
 	/* create pre-declared device nodes */
 	of_i2c_register_devices(adap);
 	i2c_acpi_install_space_handler(adap);
@@ -1618,7 +1618,6 @@ static int i2c_register_adapter(struct i2c_adapter *adap)
 	bus_for_each_drv(&i2c_bus_type, NULL, adap, __process_new_adapter);
 	mutex_unlock(&core_lock);
 
-	mutex_init(&adap->hold_lock);
 	INIT_DELAYED_WORK(&adap->unhold_work, i2c_adapter_unhold_work);
 
 	return 0;
-- 
2.41.0


From 21ae882a2114fcd0e7dbf39899e01c83adca3f18 Mon Sep 17 00:00:00 2001
From: "Arun P. Mohanan" <arun.p.m@linux.intel.com>
Date: Wed, 27 Apr 2022 10:16:47 +0530
Subject: [PATCH 5/5] i2c: mux: Add mux deselect support on timeout

Add support to deselect the mux when there is a timeout.
The mux idle_state settings will be configured on startup. In case of
MCTP it is MUX_IDLE_DISCONNECT. But when there is a timeout, mux ends
up in connected position and the devices behind the mux will appear under
different muxes connected to the same bus. This change fix the same.

Signed-off-by: Arun P. Mohanan <arun.p.m@linux.intel.com>
---
 drivers/i2c/i2c-mux.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/drivers/i2c/i2c-mux.c b/drivers/i2c/i2c-mux.c
index c5290e4062a7..4b295b8a5d4b 100644
--- a/drivers/i2c/i2c-mux.c
+++ b/drivers/i2c/i2c-mux.c
@@ -36,9 +36,11 @@ struct i2c_mux_priv {
 	u32 chan_id;
 };
 
-static void i2c_mux_hold(struct i2c_mux_core *muxc, unsigned long timeout)
+static void i2c_mux_hold(struct i2c_mux_core *muxc, u32 chan_id,
+			 unsigned long timeout)
 {
 	mutex_lock(&muxc->hold_lock);
+	muxc->holder_chan_id = chan_id;
 	schedule_delayed_work(&muxc->unhold_work, timeout);
 }
 
@@ -54,6 +56,9 @@ static void i2c_mux_unhold_work(struct work_struct *work)
 	struct i2c_mux_core *muxc = container_of(dwork, struct i2c_mux_core,
 						 unhold_work);
 
+	if (muxc->deselect)
+		muxc->deselect(muxc, muxc->holder_chan_id);
+
 	mutex_unlock(&muxc->hold_lock);
 }
 
@@ -74,7 +79,7 @@ static int __i2c_mux_master_xfer(struct i2c_adapter *adap,
 				      (u16 *)msgs[num - 1].buf);
 	if (hold_msg == I2C_HOLD_MSG_SET) {
 		timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
-		i2c_mux_hold(muxc, timeout);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
 	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
 		mutex_lock(&muxc->hold_lock);
 	}
@@ -112,7 +117,7 @@ static int i2c_mux_master_xfer(struct i2c_adapter *adap,
 				      (u16 *)msgs[num - 1].buf);
 	if (hold_msg == I2C_HOLD_MSG_SET) {
 		timeout = msecs_to_jiffies(*(u16 *)msgs[num - 1].buf);
-		i2c_mux_hold(muxc, timeout);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
 	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
 		mutex_lock(&muxc->hold_lock);
 	}
@@ -150,7 +155,7 @@ static int __i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 				      &data->word);
 	if (hold_msg == I2C_HOLD_MSG_SET) {
 		timeout = msecs_to_jiffies(data->word);
-		i2c_mux_hold(muxc, timeout);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
 	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
 		mutex_lock(&muxc->hold_lock);
 	}
@@ -189,7 +194,7 @@ static int i2c_mux_smbus_xfer(struct i2c_adapter *adap,
 				      &data->word);
 	if (hold_msg == I2C_HOLD_MSG_SET) {
 		timeout = msecs_to_jiffies(data->word);
-		i2c_mux_hold(muxc, timeout);
+		i2c_mux_hold(muxc, priv->chan_id, timeout);
 	} else if (hold_msg == I2C_HOLD_MSG_NONE) {
 		mutex_lock(&muxc->hold_lock);
 	}
-- 
2.41.0

